cpp/src/arrow/acero/unmaterialized_table.h

```diff
@@ -18,10 +18,12 @@
 #pragma once
 
 #include <optional>
+#include <unordered_set>
```

Comment:
To aid review as this branch is rebased on top of https://github.com/apache/arrow/pull/41125

This line is changed by this PR.